entities/PropertyManagementService.js

```diff
@@ -0,0 +1,199 @@
+        playerIndex
+      );
+      // multiply by roll
+      rentAmount =
+        totalRoll *
+        (hasMonopoly ? DOUBLE_UTILITY_MULTIPLIER : SINGLE_UTILITY_MULTIPLIER);
+    } else if (boardProperty.group === 'Railroad') {
+      const railroadPricing = [25, 50, 100, 200];
+      const railroadCount = this.getProperties(gameState)
+        .filter((p) => p.group === boardProperty.group)
+        .filter((p) => p.ownedBy === playerIndex).length;
+      rentAmount = railroadPricing[railroadCount - 1];
+    } else {
+      if (boardProperty.buildings > 0) {
+        rentAmount = boardProperty.multipliedRent[boardProperty.buildings - 1];
+      } else {
+        const hasMonopoly = this.hasMonopoly(
+          gameState,
+          boardProperty.group,
+          playerIndex
+        );
+
+        if (hasMonopoly) {
+          rentAmount *= 2;
+        }
+      }
+    }
+    return rentAmount;
+  }
+
+  static changeOwner(boardProperty, playerId) {
+    boardProperty.ownedBy = playerId;
+  }
+
+  static renovate(gameState, property) {
+    gameState.config.propertyConfig[
+      property.buildings !== 4 ? 'houses' : 'hotels'
+    ] -= 1;
+    property.buildings += 1;
+    WealthService.buyAsset(gameState.currentPlayer, property.houseCost);
+  }
+
+  static demolish(gameState, property) {
+    gameState.config.propertyConfig[
+      property.buildings !== 5 ? 'houses' : 'hotels'
+    ] += 1;
+    property.buildings -= 1;
+    WealthService.sellAsset(gameState.currentPlayer, property.houseCost / 2);
+  }
+
+  static toggleMortgageOnProperty(
+    gameState,
+    boardProperty,
+    player = gameState.currentPlayer
+  ) {
+    boardProperty.mortgaged = !boardProperty.mortgaged;
+    const INTEREST_RATE = 0.1;
+    const mortgageBaseCost = boardProperty.price / 2;
+
+    if (boardProperty.mortgaged) {
+      WealthService.sellAsset(player, mortgageBaseCost);
+    } else {
+      WealthService.decrement(player, mortgageBaseCost * INTEREST_RATE);
+      WealthService.buyAsset(player, mortgageBaseCost);
+    }
+  }
+
+  static hasMonopoly(gameState, propertyGroup, playerId) {
+    return this.getProperties(gameState)
+      .filter((p) => p.group === propertyGroup)
+      .every((p) => p.ownedBy === playerId);
+  }
+
+  static getAvailableManagementActions(gameState) {
+    let availableActions = ['TOGGLE_MORTGAGE'];
+
+    if (this.getRenoProperties(gameState).length > 0) {
+      availableActions.push('RENOVATE');
+    }
+
+    if (this.getDemoProperties(gameState).length > 0) {
+      availableActions.push('DEMOLISH');
+    }
+
+    availableActions.push('CANCEL');
+    return availableActions;
+  }
+
+  static getRenoProperties(gameState) {
+    const { houses, hotels, properties } = gameState.config.propertyConfig;
+    const player = gameState.currentPlayer;
+    if (houses + hotels === 0) return [];
+    const limitedHouses = houses === 0;
+    const limitedHotels = hotels === 0;
+
+    const availableRenoProps = _(properties)
```

Comment:
Just a heads up - I've used this pattern with lodash before and got some bundling errors with it
Webpack had trouble with treeshaking it because it _looks_ like it's just bringing in one thing (the monad constructor) but because it wraps it and all methods can be chained _all_ those methods are needed. Webpack failed to pull everything in and the type was missing some chained functions.
If you plan on browser-ifying this, I would consider switching this out for the explicit functions instead of chaining. You'd drop some cost, too:

- `filter` can be browser-native
- `groupBy` is an easy polyfill with `reduce` (I wrote one for Vincent this week, no joke)
- optionally `flattenDeep` can be polyfilled

You'd just need `mapValues`, then, and possibly `flattenDeep`

_**COUNTER ARGUMENT:**_ If you need performance, though, using this chaining should give you shortcut-fusion, so yeah, leave it as-is

Reply:
It's so much prettier this way :'(